hoomd/mpcd/ReverseNonequilibriumShearFlow.cc

```diff
@@ -0,0 +1,327 @@
+        }
+    }
+
+/*!
+ * \param timestep Current time step of the simulation
+ */
+void mpcd::ReverseNonequilibriumShearFlow::update(uint64_t timestep)
+    {
+    // reset momentum exchange for this step
+    m_momentum_exchange = Scalar(0);
+
+    // don't do anything if this is an empty group
+    const unsigned int n_mpcd = m_mpcd_pdata->getN();
+    if (n_mpcd == 0)
+        return;
+
+    // if slabs have changed, update them
+    if (m_update_slabs)
+        {
+        setSlabs();
+        m_update_slabs = false;
+        }
+
+    findSwapParticles();
+    swapPairMomentum();
+    }
+
+std::vector<std::string> mpcd::ReverseNonequilibriumShearFlow::getProvidedLogQuantities()
+    {
+    std::vector<std::string> ret {"rp_momentum"};
+    return ret;
+    }
+
+/*!
+ * \param quantity Name of the log quantity to get
+ * \param timestep Current time step of the simulation
+ */
+Scalar mpcd::ReverseNonequilibriumShearFlow::getLogValue(const std::string& quantity,
+                                                         uint64_t timestep)
+    {
+    if (quantity == "rp_momentum")
+        {
+        return m_momentum_exchange;
+        }
+    else
+        {
+        throw std::runtime_error("Error getting log value");
+        }
+    return Scalar(0.0);
+    }
+
+/*!
+ * Finds all particles in the "min" and "max" slab in z direction and
+ * puts them into two GPUArrays, sorted by their momentum closest to p_target in x direction.
+ */
+void mpcd::ReverseNonequilibriumShearFlow::findSwapParticles()
+    {
+    // get needed array handles. All in read mode, no particle is changed in this search function
+    ArrayHandle<Scalar4> h_vel(m_mpcd_pdata->getVelocities(),
+                               access_location::host,
+                               access_mode::read);
+    ArrayHandle<Scalar4> h_pos(m_mpcd_pdata->getPositions(),
+                               access_location::host,
+                               access_mode::read);
+
+    // sort pairs according to their sign and momentum
+    // temporary pairs to hold particles in slab
+    // first index is momentum in x dir - second index is particle tag
+    detail::ReverseNonequilibriumShearFlowSorter cmp(m_p_target);
+    std::map<Scalar, unsigned int, detail::ReverseNonequilibriumShearFlowSorter>
+        particles_in_lo_slab(cmp);
+    std::map<Scalar, unsigned int, detail::ReverseNonequilibriumShearFlowSorter>
+        particles_in_hi_slab(cmp);
+
+    // sort particles into their slab in z-direction and record momentum in x-direction and index
+    // injecting into std::map will sort them from lowest to highest momentum
+    for (unsigned int j = 0; j < m_mpcd_pdata->getN(); j++)
+        {
+        const Scalar4 vel = h_vel.data[j];
+        const Scalar momentum = vel.x * m_mpcd_pdata->getMass();
+        const Scalar z = h_pos.data[j].z;
+        if (m_lo_pos.x <= z && z < m_lo_pos.y
+            && momentum > 0) // lower slab, search for max momentum
```

Comment:
I suspect this should not be here or in the else if branch, since the sorter should handle this.
```suggestion
            ) // lower slab, search for max momentum
```